yarl/_url.py

```diff
@@ -1153,7 +1166,49 @@ def join(self, url):
         # See docs for urllib.parse.urljoin
         if not isinstance(url, URL):
             raise TypeError("url should be URL")
-        return URL(urljoin(str(self), str(url)), encoded=True)
+        other: URL = url
+        scheme = other.scheme or self.scheme
+        parts = {
+            k: getattr(self, k) for k in ["authority", "path", "query", "fragment"]
+        }
+        parts["scheme"] = scheme
+        from urllib.parse import uses_netloc as uses_authority
+        from urllib.parse import uses_relative
+
+        if scheme != self.scheme or scheme not in uses_relative:
+            return URL(str(other))
+
+        if scheme in uses_authority:
+            if other.authority:
+                parts.update(
+                    {
+                        k: getattr(other, k)
```

Comment:
Are you sure about this one? It will replace the parts with `None` if the attributes don't exist.

Reply:
Yes.
If the "other" has an authority, we use it's path/query/fragment.

c.f. 
https://github.com/python/cpython/blob/a9bb3c7b3bd8ba90de87d03cd4d614dca764b116/Lib/urllib/parse.py#L555-L557